ee/server/service/calendar.go

```diff
@@ -129,5 +191,151 @@ func (svc *Service) CalendarWebhook(ctx context.Context, eventUUID string, chann
 			return ctxerr.Wrap(ctx, err, "create or update calendar event")
 		}
 	}
+
 	return nil
 }
+
+func (svc *Service) releaseCalendarLock(ctx context.Context, eventUUID string, lockValue string) {
+	ok, err := svc.distributedLock.ReleaseLock(ctx, calendar.LockKeyPrefix+eventUUID, lockValue)
+	if err != nil {
+		level.Error(svc.logger).Log("msg", "Failed to release calendar lock", "err", err)
+	}
+	if !ok {
+		// If the lock was not released, it will expire on its own.
+		level.Warn(svc.logger).Log("msg", "Failed to release calendar lock")
+	}
+}
+
+func (svc *Service) getCalendarLock(ctx context.Context, eventUUID string, addToQueue bool) (lockValue string, reserved bool, err error) {
+	// Check if lock has been reserved, which means we can't have it.
+	reservedValue, err := svc.distributedLock.Get(ctx, calendar.ReservedLockKeyPrefix+eventUUID)
+	if err != nil {
+		return "", false, ctxerr.Wrap(ctx, err, "get calendar reserved lock")
+	}
+	reserved = reservedValue != nil
+	if reserved && !addToQueue {
+		// We flag the lock as reserved.
+		return "", reserved, nil
+	}
+	var result string
+	if !reserved {
+		// Try to acquire the lock
+		lockValue = uuid.New().String()
+		result, err = svc.distributedLock.AcquireLock(ctx, calendar.LockKeyPrefix+eventUUID, lockValue, 0)
+		if err != nil {
+			return "", false, ctxerr.Wrap(ctx, err, "acquire calendar lock")
+		}
+	}
+	if (result == "" || reserved) && addToQueue {
+		// Could not acquire lock, so we are already processing this event. In this case, we add the event to
+		// the queue (actually a set) to indicate that we need to re-process the event.
+		err = svc.distributedLock.AddToSet(ctx, calendar.QueueKey, eventUUID)
+		if err != nil {
+			return "", false, ctxerr.Wrap(ctx, err, "add calendar event to queue")
+		}
+
+		if reserved {
+			// We flag the lock as reserved.
+			return "", reserved, nil
+		}
+
+		// Try to acquire the lock again in case it was released while we were adding the event to the queue.
+		result, err = svc.distributedLock.AcquireLock(ctx, calendar.LockKeyPrefix+eventUUID, lockValue, 0)
+		if err != nil {
+			return "", false, ctxerr.Wrap(ctx, err, "acquire calendar lock again")
+		}
+
+		if result == "" {
+			// We could not acquire the lock, so we are done here.
+			return "", reserved, nil
+		}
```

Comment:
Yes, it catches the corner case where an event UUID was added to the queue, but no server actually processes the queue because all the locks have been released.